onnx/onnx_backend.h

```diff
@@ -0,0 +1,801 @@
+   *
+   * If the backend does not support FP32 computation, it must return 0.
+   */
+  uint64_t macsFP32;
+  /**
+   * Number of FP16 multiply-accumulate operations per second delivered by
+   * the backend.
+   *
+   * If the backend does not support FP16 computation, it must return 0.
+   */
+  uint64_t macsFP16;
+  /**
+   * Bandwidth, in bytes per second, of the global memory specific to the
+   * backend device.
+   */
+  uint64_t deviceMemoryBandwidth;
+  /**
+   * Bandwidth, in bytes per second, of transferring data from cacheable
+   * CPU-allocated memory to the backend device.
+   */
+  uint64_t cpuMemoryReadBandwidth;
+  /**
+   * Bandwidth, in bytes per second, of transferring data to cacheable
+   * CPU-allocated memory from the backend device.
+   */
+  uint64_t cpuMemoryWriteBandwidth;
+} onnxBackendPerformanceInfo;
+
+/* Note: the data type values match ONNX TensorProto.DataType enum */
+#define ONNX_DATATYPE_UNDEFINED 0
+#define ONNX_DATATYPE_FLOAT16 10
+#define ONNX_DATATYPE_FLOAT32 1
+#define ONNX_DATATYPE_FLOAT64 11
+#define ONNX_DATATYPE_INT8 3
+#define ONNX_DATATYPE_INT16 5
+#define ONNX_DATATYPE_INT32 6
+#define ONNX_DATATYPE_INT64 7
+#define ONNX_DATATYPE_UINT8 2
+#define ONNX_DATATYPE_UINT16 4
+#define ONNX_DATATYPE_UINT32 12
+#define ONNX_DATATYPE_UINT64 13
+
+/** Cacheable CPU memory */
+#define ONNX_MEMORY_TYPE_CPU 0
+
+#define ONNX_TENSOR_DIMS_MAX 8
+
+/**
+ * Terminates the list of auxiliary backend initialization properties passed to
+ * onnxInitBackend.
+ */
+#define ONNX_BACKEND_PROPERTY_NONE 0
+
+typedef struct onnxTensorDescriptor {
+  /**
+   * Name of the blob corresponding to this tensor in the ONNX model. The name
+   * must exactly match the ValueInfoProto.name of one of the
+   * ModelProto.graph.input or ModelProto.graph.output
+   */
+  const char* name;
+  /**
+   * Data type of the elements in the tensor.
+   *
+   * Possible values:
+   *     ONNX_DATATYPE_FLOAT16
+   *     ONNX_DATATYPE_FLOAT32
+   *     ONNX_DATATYPE_INT8
+   *     ONNX_DATATYPE_INT16
+   *     ONNX_DATATYPE_INT32
+   *     ONNX_DATATYPE_UINT8
+   *     ONNX_DATATYPE_UINT16
+   *     ONNX_DATATYPE_UINT32
+   */
+  onnxEnum dataType;
+  /**
+   * Type of memory that stores the tensor.
+   *
+   * Possible values:
+   *     ONNX_MEMORY_TYPE_CPU
+   */
+  onnxEnum memoryType;
+  /**
+   * Number of dimensions in the tensor.
+   * Must be between 0 (for a scalar) and ONNX_TENSOR_DIMS_MAX.
+   */
+  uint32_t dimensions;
+  /**
+   * Dimensions of the tensor.
+   */
+  uint64_t shape[ONNX_TENSOR_DIMS_MAX];
+  /**
+   * Pointers to tensor data.
+   */
+  onnxPointer buffer;
+} onnxTensorDescriptor;
+
+/* Function pointer declarations for dynamic loading */
+typedef uint32_t(ONNX_ABI* onnxGetNumBackendsFunction)(void);
+typedef onnxStatus(ONNX_ABI* onnxGetBackendInfoFunction)(
+    uint32_t index,
+    void* backendInfo,
+    uint32_t backendInfoVersion);
+typedef onnxStatus(ONNX_ABI* onnxGetBackendCompatibilityFunction)(
```

Comment:
If I want to (1) check backend compatibility for my whole model and then (2) run it, I have to send the protobuf once to `onnxGetBackendCompatibility`, and then to `onnxInitGraph`, forcing the backend to parse it twice. Can we combine this? For example by having a `onnxLoadGraph` which is called first, and creates  a handle that can be passed in to `onnxGetBackendCompatibility` or `onnxInitGraph`?